Create URL Deduplication Tool

maxachis opened the pull request:
#### Fixes

[#17 - Agency Table Cleaning](https://github.com/Police-Data-Accessibility-Project/data-source-identification/issues/17)

#### Description

This is a browser-based tool used to assist in resolving duplicate URLs in the PDAP Agencies Airtable. As designed, this tool will take a csv from the airtable, process it and identify duplicates, and then generate proposed fixes for the duplicates (via Google search) which the user can then, through the app, manually approve or reject. The user can then export this data as a csv to be re-inserted into the Airtable.

#### Testing

Upon downloading, the README file should contain all necessary instructions for running the application. Following the instructions should lead to the production of a csv file called "export.csv" which should be able to be inserted into the Airtable. 

#### Docs

No Docs at this time.

maxachis committed:
Create repository

maxachis committed:
Create repository

maxachis committed:
Update database

maxachis committed:
Update database

maxachis committed:
Create DatabaseManager

maxachis committed:
Create DatabaseManager

maxachis committed:
Flesh out identify duplicates

maxachis committed:
Flesh out identify duplicates

maxachis committed:
Create initial_data_load script

maxachis committed:
Create initial_data_load script

maxachis committed:
Create empty util script for later

maxachis committed:
Create empty util script for later

maxachis committed:
Create unfinished requirements.txt

maxachis committed:
Create unfinished requirements.txt

maxachis committed:
Flash out pipeline with still-incomplete draft of pipeline.py

maxachis committed:
Flash out pipeline with still-incomplete draft of pipeline.py

maxachis committed:
Bootstrapify HTML

maxachis committed:
Bootstrapify HTML

maxachis committed:
Bootstrapify HTML, add preview options.

maxachis committed:
Bootstrapify HTML, add preview options.

maxachis committed:
Rename data model to reflect database

maxachis committed:
Rename data model to reflect database

maxachis committed:
Remove unused stylesheet

maxachis committed:
Remove unused stylesheet

maxachis committed:
Simplify pipeline

maxachis committed:
Simplify pipeline

maxachis committed:
Update database

maxachis committed:
Update database

maxachis committed:
Replace HTTPError with RuntimeError

maxachis committed:
Replace HTTPError with RuntimeError

maxachis committed:
Refine Deduplicator
-- Now directly updates database after every successful entry
-- Includes conversion of state acronym to state name
-- Includes in-place print logging

maxachis committed:
Refine Deduplicator
-- Now directly updates database after every successful entry
-- Includes conversion of state acronym to state name
-- Includes in-place print logging

maxachis committed:
Refine get_entries_without_proposed_url

maxachis committed:
Refine get_entries_without_proposed_url

maxachis committed:
Update Database Manager
-- Modify to account for new table structure
-- Add reset_database method
-- Add get_completed_entries method

maxachis committed:
Update Database Manager
-- Modify to account for new table structure
-- Add reset_database method
-- Add get_completed_entries method

maxachis committed:
Modify to account for change in database structure

maxachis committed:
Modify to account for change in database structure

maxachis committed:
Create README

maxachis committed:
Create README

maxachis committed:
Update requirements.txt

maxachis committed:
Update requirements.txt

maxachis committed:
Modify to account for change in database structure

maxachis committed:
Modify to account for change in database structure

maxachis committed:
Add functionality so that blank urls are skipped

maxachis committed:
Add functionality so that blank urls are skipped

maxachis committed:
Add comments and modify to account for change in database structure

maxachis committed:
Add comments and modify to account for change in database structure

maxachis committed:
Create PDAPInterface (not yet functional)

maxachis committed:
Create PDAPInterface (not yet functional)

josh-chamberlain commented:
@maxachis sorry for the delay, now that I'm catching up I'll test this for usability soon

josh-chamberlain commented on README.md:
```diff
@@ -0,0 +1,39 @@
+
+Pulls data from [PDAP Agencies Airtable](https://airtable.com/app473MWXVJVaD7Es/shr43ihbyM8DDkKx4/tblpnd3ei5SlibcCX).
+
+To utilize this tool, either utilize the PDAP Interface (not currently completed) or download a csv of the Agencies data and run initial_data_load.py to load the data into the database.
```
what do you mean by "PDAP Interface" here?

maxachis replied:
In that case I'm referring to the [PDAPInterface](https://github.com/maxachis/data-cleaning/blob/main/data_processing/PDAPInterface.py) class, which should streamline some of the functionality.

josh-chamberlain replied:
Is this intended to be an interface for interacting with our own API, updating things and the like?

josh-chamberlain commented on requirements.txt:
```diff
@@ -0,0 +1,9 @@
+pyarrow~=14.02
```
should be `14.0.2` I believe

maxachis replied:
Corrected in latest commit.

josh-chamberlain commented on README.md:
```diff
@@ -0,0 +1,39 @@
+Pulls data from [PDAP Agencies Airtable](https://airtable.com/app473MWXVJVaD7Es/shr43ihbyM8DDkKx4/tblpnd3ei5SlibcCX).
+
+To utilize this tool, either utilize the PDAP Interface (not currently completed) or download a csv of the Agencies data and run initial_data_load.py to load the data into the database.
+
```
before this step I had to `pip install -r requirements.txt`

maxachis replied:
Added that to the README, per this comment.

josh-chamberlain commented on README.md:
````diff
@@ -0,0 +1,39 @@
+And then run pipeline.py to pull data from the web and update the database.
+
+```bash
+python data_processing/pipeline.py
````
i had to add `google-api-python-client` to the requirements.txt here. I struggled a bit with polars install because of cache stuff—always worth telling people to use a `venv`, imo, because the install is more predictable

josh-chamberlain replied:
This is where I'm stuck

```
Traceback (most recent call last):
  File "/Users/joshchamberlain/dev/pdap/data-cleaning/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1970, in _exec_single_context
    self.dialect.do_execute(
  File "/Users/joshchamberlain/dev/pdap/data-cleaning/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: no such table: agencies

```

maxachis replied:
@josh-chamberlain I attempted to address this in the most recent version of my commits. Also other details, which I'll describe in my comment below.

josh-chamberlain replied:
I did get a better error: 

```
RuntimeError: Custom search API key and CSE ID must be set in .env file
```

and I was able to resolve it with the README.

maxachis committed:
Correct typo in class name

maxachis committed:
Add instruction to install requirements.

maxachis committed:
Correct typo in requirements.txt

maxachis committed:
Add additional context to source retrieval instruction

maxachis committed:
Move config to app folder

maxachis committed:
Modify imports

maxachis committed:
Further refine Readme
-- Add additional clarifications to source retrieval step
-- Move environment instructions into README at pipeline step

maxachis committed:
Add headers to README

maxachis committed:
Add clarification to "Data Processing" step

maxachis committed:
Add check if `.env` entries are set

maxachis committed:
Refine template and static folder addressing for flexibility
Remove unused import

maxachis committed:
Add venv instructions

maxachis committed:
Add clarifying comment

maxachis committed:
Update database path setting

maxachis committed:
Add database existence checking functionality.

maxachis committed:
Add google-api-python-client to requirements.txt

maxachis committed:
Update

maxachis committed:
Create .gitignore with .env file

maxachis committed:
Corrected dreaded imports

maxachis committed:
Update error checking

maxachis commented:
I made a number of updates to the repo:

- Updated and corrected requirements.txt
- Add checkers for common errors
- Updated how certain files, such as the database, are being referenced -- these should be more resilient, or there should be errors that are more descriptive.
- Updated the readme for better clarity (including adding recommendations for venv).

josh-chamberlain commented:
Hey, I got it running! Nice.

A few thoughts:

- For clarity: instead of showing one agency name, we should probably show one URL and list the agencies which claim the URL. the user can select `URL is correct` for one agency, and decide the correct URL for any other agencies presented as demonstrated by this workflow.

- let's say the suggestion is close, but wrong. Say it gives "https://www.allencountysheriff.org/county-government", but I just want to remove the /county-government part. If I paste whatever URL I want in the `Possible Correct URL` field, and click "accept", will my override by captured?

- in this case, the old URL and suggested URL were nearly the same, separated by `https`: <img width="1137" alt="Screen Shot 2024-03-26 at 3 05 07 PM" src="https://github.com/Police-Data-Accessibility-Project/data-cleaning/assets/30379833/85518099-09e0-4675-8e97-47d0f03aaef2">

- `python data_processing/export.py` doesn't seem to exist 👀

mbodeantor commented on data_processing/Deduplicator.py:
```diff
@@ -0,0 +1,160 @@
+from GoogleSearcher import GoogleSearcher
+from data_processing.DatabaseManager import DatabaseManager
+
+states = {
```
can use states table here too

mbodeantor reviewed (commented):
Could use more clarity around what is expected in agencies.csv in absence of the completed get_agencies_data function

maxachis commented:
@josh-chamberlain I'll happily make the adjustments given by you and @mbodeantor.

That said, I'm thinking it may be useful to modify how the data is being pulled so that it pulls directly from the database via the API, rather than the option to pull from either the Airtable or a csv from the Airtable. My reasons are:

1. Pulling via the API will help ensure data is up-to-date and require less work on the part of a volunteer, who will just need to make sure they have data access
2. This will give me an opportunity to familiarize myself with setting up the API interface. 
3. To use the `states` table in the database, as @mbodeantor recommends, it'll be easier if I do it through this method.

Let me know your thoughts.

josh-chamberlain commented:
@maxachis yes, pulling via the API is probably preferred. I think the CSV strategy still makes sense as a workable solution, but proceed as you wish there.
